util/codegen/codegen_test.go

```diff
@@ -174,3 +179,103 @@ func TestGenericContainsPointers(t *testing.T) {
 		})
 	}
 }
+
+func TestAssertStructUnchanged(t *testing.T) {
+	type args struct {
+		t      *types.Struct
+		tname  string
+		params *types.TypeParamList
+		ctx    string
+		it     *ImportTracker
+	}
+	cfg := &packages.Config{
+		Mode:  packages.NeedTypes | packages.NeedTypesInfo | packages.NeedSyntax | packages.NeedName,
+		Tests: true,
+	}
+
+	rootDir := "./codegen_test_pkgs"
+	pkg_map := make(map[string]*types.Package)
+
+	// load packages recursively
+	err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			pkgs, _ := packages.Load(cfg, "./"+path)
+			for _, pkg := range pkgs {
+				if len(pkg.Errors) == 0 {
+					pkg_map[pkg.Name] = pkg.Types
+				}
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	tests := []struct {
+		name string
+		args args
+		want []byte
+	}{
+		{
+			name: "t1-internally_defined",
+			args: args{
+				t: types.NewStruct([]*types.Var{
```

Comment:
Can you please either use both packages and types from `pkg_map`, or define both in the code as you do with types?

That is, either use something like `pkg_map["t1"].Scope().Lookup("T1").Type().Underlying().(*types.Struct)` as the type, or `p1 := types.NewPackage("t1", "t1")` to create a package in memory and use `p1` in place of `pkg_map["t1"]`.

I personally prefer keeping your type definitions as is and defining packages instead of loading them. It’s less code and helps keep tests self-contained. But either is fine.

Reply:
You are right, it's much cleaner. It's now self-contained.